airflow_prometheus_exporter/prometheus_exporter.py

```diff
@@ -381,30 +383,71 @@ def get_num_queued_tasks():
         )
 
 
+seconds_per_unit = {"s": 1, "m": 60, "h": 3600, "d": 86400, "w": 604800}
+sla_interval_pattern = re.compile("[ ]*([0-9]+).*(s|m|h|d|w)[ ]*")
+
+
+def convert_to_seconds(sla_interval):
+    match = re.match(sla_interval_pattern, sla_interval)
+    return int(match.group(1)) * seconds_per_unit[match.group(2)]
+
+
+def sla_check(sla_interval, sla_time, max_execution_date, cadence, execution_dates):
+    utc_datetime = pytz.timezone("UTC").localize(datetime.datetime.utcnow())
+    if sla_time:
+        local_datetime = utc_datetime.astimezone(pytz.timezone(TIMEZONE_LA))
+        sla_datetime = pytz.timezone(TIMEZONE_LA).localize(
+            datetime.datetime.combine(
+                local_datetime.date(), datetime.time(int(sla_time))
+            )
+        )
+    else:
+        sla_datetime = utc_datetime
+
+    interval_in_second = convert_to_seconds(sla_interval)
+    checkpoint = sla_datetime - datetime.timedelta(seconds=interval_in_second)
+    if utc_datetime >= sla_datetime and max_execution_date < checkpoint:
+        return True
+
+    if cadence != "triggered":
+        # Check the state of previous run before sla_time.
+        # To detect consecutive failed scenario.
+        # Filter out triggered DAGs e.g. PPD
+        for record in execution_dates:
+            if record["execution_date"] > checkpoint:
+                continue
+            if record["state"] != "success":
+                return True
+            break
+
+    return False
+
+
 def get_sla_miss_dags():
     min_date_to_filter = pendulum.now(TIMEZONE).subtract(days=RETENTION_TIME)
     with session_scope(Session) as session:
-        max_execution_dt_query = (
+        execution_dt_query = (
             session.query(
                 DagRun.dag_id,
                 DagModel.schedule_interval,
-                func.max(DagRun.execution_date).label("max_execution_date"),
+                DagRun.execution_date,
+                DagRun.state,
             )
             .join(DagModel, DagModel.dag_id == DagRun.dag_id)
             .filter(
-                DagModel.is_active == True,  # noqa
+                DagModel.is_active == True,
                 DagModel.is_paused == False,
-                DagRun.state == State.SUCCESS,
                 DagRun.execution_date > min_date_to_filter,
             )
-            .group_by(DagRun.dag_id, DagModel.schedule_interval)
+            .order_by(DagRun.execution_date.desc())
             .subquery()
         )
```

Comment:
Need to get all execution dates with states. In case we need to find the latest state outside of sla window. You find the same in task.